File: src/typescript/src/algorithms/kmeans.ts
import { Samples } from "../preprocessing.js";

export interface CompressionResult {
  compressed: Samples;
  labels: Uint32Array;
  centers: Float32Array;
  compactness: number;
}

export interface KMeansConfig {
  maxIterations?: number;
  tolerance?: number;
}

export class KMeans {
  private readonly maxIterations: number;
  private readonly tolerance: number;

  constructor({ maxIterations = 100, tolerance = 1e-4 }: KMeansConfig = {}) {
    this.maxIterations = maxIterations;
    this.tolerance = tolerance;
  }

  compress(samples: Samples, clusters: number): CompressionResult {
    if (samples.cols !== 3) {
      throw new Error("samples must have 3 columns");
    }
    if (clusters < 2 || clusters > 256) {
      throw new Error("clusters must be in [2, 256]");
    }
    if (samples.rows < clusters) {
      throw new Error("sample count must be >= clusters");
    }

    const centers = this.initializeCenters(samples, clusters);
    const labels = new Uint32Array(samples.rows);

    for (let iter = 0; iter < this.maxIterations; iter += 1) {
      this.assignClusters(samples, centers, labels);
      const newCenters = this.updateCenters(samples, labels, clusters);

      if (this.hasConverged(centers, newCenters)) {
        centers.set(newCenters);
        break;
      }

      centers.set(newCenters);
    }

    this.assignClusters(samples, centers, labels);
    const compressed = this.buildCompressedSamples(samples, centers, labels);
    const compactness = this.computeCompactness(samples, compressed);

    return {
      compressed,
      labels,
      centers,
      compactness,
    };
  }

  private initializeCenters(samples: Samples, clusters: number): Float32Array {
    const centers = new Float32Array(clusters * samples.cols);
    const indices = shuffle(range(samples.rows));

    for (let i = 0; i < clusters; i += 1) {
      const sampleIdx = indices[i] * samples.cols;
      const centerIdx = i * samples.cols;
      centers[centerIdx] = samples.data[sampleIdx];
      centers[centerIdx + 1] = samples.data[sampleIdx + 1];
      centers[centerIdx + 2] = samples.data[sampleIdx + 2];
    }

    return centers;
  }

  private assignClusters(samples: Samples, centers: Float32Array, labels: Uint32Array): void {
    const { rows, cols } = samples;
    const clusterCount = centers.length / cols;

    for (let i = 0; i < rows; i += 1) {
      const sampleIdx = i * cols;
      let bestCluster = 0;
      let bestDist = Number.POSITIVE_INFINITY;

      for (let c = 0; c < clusterCount; c += 1) {
        const centerIdx = c * cols;
        const dist = squaredDistance(samples.data, centers, sampleIdx, centerIdx);

        if (dist < bestDist) {
          bestDist = dist;
          bestCluster = c;
        }
      }

      labels[i] = bestCluster;
    }
  }

  private updateCenters(samples: Samples, labels: Uint32Array, clusters: number): Float32Array {
    const { rows, cols } = samples;
    const centers = new Float32Array(clusters * cols);
    const counts = new Uint32Array(clusters);

    for (let i = 0; i < rows; i += 1) {
      const cluster = labels[i];
      const sampleIdx = i * cols;
      const centerIdx = cluster * cols;
      centers[centerIdx] += samples.data[sampleIdx];
      centers[centerIdx + 1] += samples.data[sampleIdx + 1];
      centers[centerIdx + 2] += samples.data[sampleIdx + 2];
      counts[cluster] += 1;
    }

    for (let c = 0; c < clusters; c += 1) {
      const count = counts[c];
      const centerIdx = c * cols;
      if (count > 0) {
        centers[centerIdx] /= count;
        centers[centerIdx + 1] /= count;
        centers[centerIdx + 2] /= count;
      }
    }

    return centers;
  }

  private hasConverged(oldCenters: Float32Array, newCenters: Float32Array): boolean {
    let maxShift = 0;
    for (let i = 0; i < oldCenters.length; i += 3) {
      const shift = Math.sqrt(
        squaredDistance(oldCenters, newCenters, i, i),
      );
      if (shift > maxShift) {
        maxShift = shift;
      }
    }
    return maxShift <= this.tolerance;
  }

  private buildCompressedSamples(
    samples: Samples,
    centers: Float32Array,
    labels: Uint32Array,
  ): Samples {
    const { rows, cols } = samples;
    const data = new Float32Array(rows * cols);

    for (let i = 0; i < rows; i += 1) {
      const cluster = labels[i];
      const sampleIdx = i * cols;
      const centerIdx = cluster * cols;
      data[sampleIdx] = centers[centerIdx];
      data[sampleIdx + 1] = centers[centerIdx + 1];
      data[sampleIdx + 2] = centers[centerIdx + 2];
    }

    return { data, rows, cols };
  }

  private computeCompactness(original: Samples, compressed: Samples): number {
    let sum = 0;
    for (let i = 0; i < original.data.length; i += 1) {
      const diff = original.data[i] - compressed.data[i];
      sum += diff * diff;
    }
    return sum;
  }
}

function squaredDistance(
  a: Float32Array,
  b: Float32Array,
  aIdx: number,
  bIdx: number,
): number {
  const dx = a[aIdx] - b[bIdx];
  const dy = a[aIdx + 1] - b[bIdx + 1];
  const dz = a[aIdx + 2] - b[bIdx + 2];
  return dx * dx + dy * dy + dz * dz;
}

function range(count: number): number[] {
  return Array.from({ length: count }, (_, idx) => idx);
}

function shuffle<T>(array: T[]): T[] {
  const result = array.slice();
  for (let i = result.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    [result[i], result[j]] = [result[j], result[i]];
  }
  return result;
}

File: src/rust/src/algorithms/kmeans.rs
use anyhow::{ensure, Result};
use ndarray::{Array2, ArrayView1};
use rand::seq::SliceRandom;
use rand::thread_rng;

#[derive(Debug)]
pub struct CompressionResult {
    pub compressed_samples: Array2<f32>,
    pub labels: Vec<usize>,
    pub centers: Array2<f32>,
    pub compactness: f64,
}

#[derive(Debug, Clone, Copy)]
pub struct Config {
    pub max_iter: usize,
    pub tolerance: f32,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            max_iter: 100,
            tolerance: 1e-4,
        }
    }
}

#[derive(Debug)]
pub struct KMeans {
    config: Config,
}

impl KMeans {
    pub fn new(config: Config) -> Self {
        Self { config }
    }

    pub fn compress(&self, samples: &Array2<f32>, clusters: usize) -> Result<CompressionResult> {
        ensure!(samples.ncols() == 3, "expected 3 columns (RGB)");
        ensure!(clusters >= 2, "clusters must be >= 2");
        ensure!(samples.nrows() >= clusters, "sample count must be >= clusters");

        let mut rng = thread_rng();
        let mut centers = initialize_centers(samples, clusters, &mut rng);
        let mut labels = vec![0usize; samples.nrows()];

        for _ in 0..self.config.max_iter {
            assign_clusters(samples, &centers, &mut labels);
            let new_centers = update_centers(samples, &labels, clusters);

            if has_converged(&centers, &new_centers, self.config.tolerance) {
                centers = new_centers;
                break;
            }

            centers = new_centers;
        }

        assign_clusters(samples, &centers, &mut labels);
        let compressed = build_compressed_samples(&centers, &labels);
        let compactness = compute_compactness(samples, &compressed);

        Ok(CompressionResult {
            compressed_samples: compressed,
            labels,
            centers,
            compactness,
        })
    }
}

impl Default for KMeans {
    fn default() -> Self {
        Self::new(Config::default())
    }
}

fn initialize_centers(
    samples: &Array2<f32>,
    clusters: usize,
    rng: &mut rand::rngs::ThreadRng,
) -> Array2<f32> {
    let mut indices: Vec<_> = (0..samples.nrows()).collect();
    indices.shuffle(rng);
    let selected = &indices[..clusters];

    let mut centers = Array2::<f32>::zeros((clusters, samples.ncols()));
    for (center_idx, &sample_idx) in selected.iter().enumerate() {
        centers
            .row_mut(center_idx)
            .assign(&samples.row(sample_idx));
    }
    centers
}

fn assign_clusters(samples: &Array2<f32>, centers: &Array2<f32>, labels: &mut [usize]) {
    for (i, sample) in samples.outer_iter().enumerate() {
        let mut best_cluster = 0usize;
        let mut best_dist = f32::MAX;

        for (cluster_idx, center) in centers.outer_iter().enumerate() {
            let dist = squared_distance(&sample, &center);
            if dist < best_dist {
                best_dist = dist;
                best_cluster = cluster_idx;
            }
        }
        labels[i] = best_cluster;
    }
}

fn update_centers(samples: &Array2<f32>, labels: &[usize], clusters: usize) -> Array2<f32> {
    let mut centers = Array2::<f32>::zeros((clusters, samples.ncols()));
    let mut counts = vec![0usize; clusters];

    for (sample, &label) in samples.outer_iter().zip(labels.iter()) {
        let mut row = centers.row_mut(label);
        row += &sample;
        counts[label] += 1;
    }

    for (mut center, count) in centers.outer_iter_mut().zip(counts.iter()) {
        if *count > 0 {
            center.map_inplace(|x| *x /= *count as f32);
        }
    }

    centers
}

fn has_converged(old: &Array2<f32>, new: &Array2<f32>, tolerance: f32) -> bool {
    let mut max_shift = 0f32;
    for (old_center, new_center) in old.outer_iter().zip(new.outer_iter()) {
        let shift = squared_distance(&old_center, &new_center).sqrt();
        if shift > max_shift {
            max_shift = shift;
        }
    }
    max_shift <= tolerance
}

fn build_compressed_samples(centers: &Array2<f32>, labels: &[usize]) -> Array2<f32> {
    let mut output = Array2::<f32>::zeros((labels.len(), centers.ncols()));
    for (i, &label) in labels.iter().enumerate() {
        output.row_mut(i).assign(&centers.row(label));
    }
    output
}

fn compute_compactness(samples: &Array2<f32>, compressed: &Array2<f32>) -> f64 {
    samples
        .outer_iter()
        .zip(compressed.outer_iter())
        .map(|(orig, comp)| squared_distance(&orig, &comp) as f64)
        .sum()
}

fn squared_distance(a: &ArrayView1<f32>, b: &ArrayView1<f32>) -> f32 {
    a.iter()
        .zip(b.iter())
        .map(|(x, y)| (x - y).powi(2))
        .sum()
}
